### PR TITLE
Fix parallel builds

### PR DESCRIPTION
**Pull request by rhaschke**
This PR addresses an issue originally raised in https://github.com/catkin/catkin_tools/issues/422, but which is actually related to the underlying catkin cmake files.
* 1dbc0ad addresses the original issue of cmake's `file RENAME` only working within a filesystem. It fails when trying to move a file _across_ file systems. However, probably the operation within a single filesystem is what makes the operation atomic in cmake.
* To counteract this regression, 6787656 also adds a `file LOCK` on the destination folder as suggested by @guihomework in https://github.com/catkin/catkin_tools/issues/422#issuecomment-414774603. This additionally avoids issues of several parallel catkin cmake processes writing the same environment files simultaneously (which is a real issue for our jenkins setup deploying directly into a shared install-space.)

### CODE REVIEW

File: cmake/atomic_configure_file.cmake [hunk 1]
```diff
@@ -1,9 +1,9 @@
...
+  get_filename_component(atomic_file ${output} NAME)
+  set(atomic_file "${CMAKE_BINARY_DIR}/${atomic_file}")
   configure_file("${input}" "${atomic_file}" ${ARGN})
   get_filename_component(output_path ${output} PATH)
-  if(NOT EXISTS ${output_path})
-    file(MAKE_DIRECTORY ${output_path})
-  endif()
-  file(RENAME "${atomic_file}" "${output}")
+  # sync multiple catkin cmake processes writing to that location
+  file(LOCK "${output_path}" DIRECTORY GUARD FUNCTION)
```
Comment (dirk-thomas) [inline cmake/atomic_configure_file.cmake]: This signature is only available as of CMake 3.2. Please use it only conditionally.

While this ensures mutually exclusive access in CMake are any of the files generated by this function accessed in parallel from non-CMake processes?

**Comment by rhaschke** [comment]
@dirk-thomas Valid point. Fixed.

File: cmake/atomic_configure_file.cmake [hunk 2]
```diff
@@ -1,9 +1,11 @@
 function(atomic_configure_file input output)
-  set(atomic_file "${CMAKE_BINARY_DIR}/atomic_configure_file")
+  get_filename_component(atomic_file ${output} NAME)
+  set(atomic_file "${CMAKE_BINARY_DIR}/${atomic_file}")
   configure_file("${input}" "${atomic_file}" ${ARGN})
   get_filename_component(output_path ${output} PATH)
-  if(NOT EXISTS ${output_path})
-    file(MAKE_DIRECTORY ${output_path})
+  # sync multiple catkin cmake processes writing to that location
+  if (CMAKE_VERSION VERSION_GREATER 3.2)
```
Comment (dirk-thomas) [inline cmake/atomic_configure_file.cmake]: The condition evaluates to `FALSE` for CMake 3.2 for which it should be `TRUE`.

File: cmake/atomic_configure_file.cmake [hunk 3]
```diff
@@ -1,9 +1,11 @@
 function(atomic_configure_file input output)
-  set(atomic_file "${CMAKE_BINARY_DIR}/atomic_configure_file")
+  get_filename_component(atomic_file ${output} NAME)
+  set(atomic_file "${CMAKE_BINARY_DIR}/${atomic_file}")
```
Comment (dirk-thomas) [inline cmake/atomic_configure_file.cmake]: To reduce the chance of collisions if the basename of `output` is the same as an existing file in the binary dir please keep `atomic_configure_file` as a prefix to the name: `set(atomic_file "${CMAKE_BINARY_DIR}/atomic_configure_file_${atomic_file}")`

File: cmake/atomic_configure_file.cmake (hunk 2 above)
Comment (rhaschke) [reply cmake/atomic_configure_file.cmake]: Fixed.

File: cmake/atomic_configure_file.cmake (hunk 3 above)
Comment (rhaschke) [reply cmake/atomic_configure_file.cmake]: This was changed intentionally. Using a different name than the final one, the copying needs two steps: first copying and then renaming the file at the destination location. (Remember that COPY doesn't allow for a simultaneous rename).
For me, this contradicts the atomic idea.

Comment (dirk-thomas) [reply cmake/atomic_configure_file.cmake]: Good point - I didn't consider that. Can you place the file in a subdirectory `atomic_configure_file` then to minimize the chance of collisions.

**Comment by dirk-thomas** [comment]
Thank you for the patch and for iterating on it.

**Pull request merged**